File: python/dnd-character/dnd_character.py
import heapq
import random

class Character:
    def __init__(self):
        # Initialize attributes with random ability scores
        self.strength = self.ability()
        self.dexterity = self.ability()
        self.constitution = self.ability()
        self.intelligence = self.ability()
        self.wisdom = self.ability()
        self.charisma = self.ability()
        
        # Calculate hitpoints based on constitution modifier
        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        # Generate an ability score by rolling four 6-sided dice and summing the top three values
        return sum(heapq.nlargest(3, random.choices(range(1, 7), k=4)))

def modifier(constitution):
    # Calculate the ability modifier based on the constitution score
    return (constitution - 10) // 2

File: javascript/dnd-character/dnd-character.js
export const abilityModifier = (constitution) => {
  return Math.floor((constitution - 10) / 2); // Calculate and return the ability modifier for constitution
};

export class Character {
  static rollAbility() {
    const rolls = Array.from(
      { length: 4 },
      () => Math.floor(Math.random() * 6) + 1,
    ); // Generate 4 random dice rolls
    rolls.sort((a, b) => b - a); // Sort the rolls in descending order
    return rolls.slice(0, 3).reduce((acc, curr) => acc + curr, 0); // Return the sum of the three highest rolls
  }

  // Getters for each ability using rollAbility() method
  get strength() {
    return Character.rollAbility();
  }

  get dexterity() {
    return Character.rollAbility();
  }

  get constitution() {
    return Character.rollAbility();
  }

  get intelligence() {
    return Character.rollAbility();
  }

  get wisdom() {
    return Character.rollAbility();
  }

  get charisma() {
    return Character.rollAbility();
  }

  // Getter for hitpoints based on constitution modifier
  get hitpoints() {
    const conModifier = abilityModifier(this.constitution); // Get constitution modifier
    return 10 + conModifier;
  }
}
